<a href="https://colab.research.google.com/github/piiop/neural-network-challenge-2/blob/main/attrition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Part 1: Preprocessing

In [102]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [103]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [104]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]


In [105]:
# Create a list of at least 10 column names to use as X data
column_list = ['Age','EnvironmentSatisfaction', 'HourlyRate',
               'JobInvolvement','JobLevel','JobSatisfaction',
               'NumCompaniesWorked','YearsAtCompany','YearsInCurrentRole',
               'YearsWithCurrManager']


# Create X_df using your selected columns
X_df = attrition_df[column_list]

# Show the data types for X_df
X_df.dtypes


Age                        int64
EnvironmentSatisfaction    int64
HourlyRate                 int64
JobInvolvement             int64
JobLevel                   int64
JobSatisfaction            int64
NumCompaniesWorked         int64
YearsAtCompany             int64
YearsInCurrentRole         int64
YearsWithCurrManager       int64
dtype: object

In [106]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=199)


In [107]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary

# All training data already int64

In [108]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
scaler.fit(X_train)


# Scale the training and testing data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [109]:
# Create a OneHotEncoder for the Department column

dept_encoder = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
dept_encoder.fit(y_train['Department'].values.reshape(-1, 1))

# Create two new variables by applying the encoder
# to the training and testing data
dept_names = dept_encoder.get_feature_names_out(['Department'])
dept_train_encoded = pd.DataFrame(dept_encoder.transform(y_train['Department'].values.reshape(-1,1)), columns=dept_names)
dept_test_encoded = pd.DataFrame(dept_encoder.transform(y_test['Department'].values.reshape(-1,1)), columns=dept_names)

display(dept_train_encoded)

Department_Human Resources  Department_Research & Development  \
0                            1.0                                0.0   
1                            0.0                                1.0   
2                            0.0                                0.0   
3                            0.0                                1.0   
4                            0.0                                1.0   
...                          ...                                ...   
1097                         0.0                                1.0   
1098                         0.0                                1.0   
1099                         1.0                                0.0   
1100                         0.0                                0.0   
1101                         0.0                                1.0   

      Department_Sales  
0                  0.0  
1                  0.0  
2                  1.0  
3                  0.0  
4                  0.0  
...                ...  
1097               0.0  
1098               0.0  
1099               0.0  
1100               1.0  
1101               0.0  

[1102 rows x 3 columns]

In [110]:

# Create a OneHotEncoder for the Attrition column
attr_encoder = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
attr_encoder.fit(y_train['Attrition'].values.reshape(-1, 1))

# Create two new variables by applying the encoder
# to the training and testing data
attr_names = attr_encoder.get_feature_names_out(['Attrition'])
attr_train_encoded = pd.DataFrame(attr_encoder.transform(y_train['Attrition'].values.reshape(-1,1)), columns=attr_names)
attr_test_encoded = pd.DataFrame(attr_encoder.transform(y_test['Attrition'].values.reshape(-1,1)), columns=attr_names)
display(attr_train_encoded)

Attrition_No  Attrition_Yes
0              1.0            0.0
1              0.0            1.0
2              1.0            0.0
3              1.0            0.0
4              1.0            0.0
...            ...            ...
1097           1.0            0.0
1098           1.0            0.0
1099           0.0            1.0
1100           1.0            0.0
1101           0.0            1.0

[1102 rows x 2 columns]

## Create, Compile, and Train the Model

In [111]:
# Find the number of columns in the X training data
columns = X_train_scaled.shape[1]

# Create the input layer
input_layer = layers.Input(shape=(columns,))

# Create at least two shared layers
shared1 = layers.Dense(units=64, activation='relu')(input_layer)
shared2 = layers.Dense(units=128, activation='relu')(shared1)


In [112]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
department_hidden = layers.Dense(units=32, activation='relu')(shared2)

# Create the output layer
department_output = layers.Dense(3, activation='softmax',name='department_output')(department_hidden)


In [113]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden = layers.Dense(units=32, activation='relu')(shared2)

# Create the output layer
attrition_output = layers.Dense(2, activation='sigmoid',name='attrition_output')(attrition_hidden)


In [114]:
# Create the model
model = Model(inputs=input_layer,
              outputs=[department_output, attrition_output])

# Compile the model
model.compile(optimizer='adam',
              loss={'department_output' : 'categorical_crossentropy',
                    'attrition_output' : 'binary_crossentropy'},
              metrics={'department_output' : ['accuracy'],
                    'attrition_output' : ['accuracy']})

# Summarize the model
model.summary()

Model: "functional_11"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_layer_6             │ (None, 10)             │              0 │ -                      │
│ (InputLayer)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_23 (Dense)          │ (None, 64)             │            704 │ input_layer_6[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_24 (Dense)          │ (None, 128)            │          8,320 │ dense_23[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_25 (Dense)          │ (None, 32)             │          4,128 │ dense_24[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_26 (Dense)          │ (None, 32)             │          4,128 │ dense_24[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_output (Dense) │ (None, 3)              │             99 │ dense_25[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_output (Dense)  │ (None, 2)              │             66 │ dense_26[0][0]         │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 17,445 (68.14 KB)

 Trainable params: 17,445 (68.14 KB)

 Non-trainable params: 0 (0.00 B)

In [115]:
# Train the model
model.fit(X_train_scaled,
        {
        'department_output': dept_train_encoded,
        'attrition_output': attr_train_encoded
        },
          epochs=100,
          batch_size=32,
          )


Epoch 1/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - attrition_output_accuracy: 0.8537 - department_output_accuracy: 0.6516 - loss: 1.3730
Epoch 2/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8315 - department_output_accuracy: 0.6529 - loss: 1.2003
Epoch 3/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8461 - department_output_accuracy: 0.6363 - loss: 1.1651
Epoch 4/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8503 - department_output_accuracy: 0.6405 - loss: 1.1468
Epoch 5/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8556 - department_output_accuracy: 0.6363 - loss: 1.1398
Epoch 6/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8453 - department_output_accuracy: 0.6361 - loss: 1.1274
Epoch 7/100
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8622 - department_output_accuracy: 0.6552 - loss: 1.0636
Epoch 8/100
35/35 ━━━━━━━━━

In [116]:
# Evaluate the model with the testing data
results = model.evaluate(X_test_scaled,
                                    {
                                        'department_output': dept_test_encoded,
                                        'attrition_output': attr_test_encoded
                                    },
                                    verbose=1)

# Print the evaluation results
print(results)

12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.7902 - department_output_accuracy: 0.5183 - loss: 3.7981  
[3.424253463745117, 0.7989130616188049, 0.570652186870575]


In [118]:
# Print the accuracy for both department and attrition
department_accuracy = results[2]
attrition_accuracy = results[1]

# Print the results
print(f"Department predictions accuracy: {department_accuracy}")
print(f"Attrition predictions accuracy: {attrition_accuracy}")

Department predictions accuracy: 0.570652186870575
Attrition predictions accuracy: 0.7989130616188049


In [121]:
print(attrition_df['Attrition'].value_counts(normalize=True))
print(attrition_df['Department'].value_counts(normalize=True))

Attrition
No     0.838776
Yes    0.161224
Name: proportion, dtype: float64
Department
Research & Development    0.653741
Sales                     0.303401
Human Resources           0.042857
Name: proportion, dtype: float64


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. Accuracy is not the best method to use. The accuracy simply does not provide enough information to evaluate the performance of the model, especially with imbalanced classes. Using a confusion matrix would give us a better view of the specifics of the models performance. Precision and Recall especially might be important to look at, along with F1 score and ROC-AUC curve or MCC.

2. For the Department output layer I used softmax due it being a multi-class classification with the outcomes are mutually exlusive. An employee can only work in one department at one time. For Attrition I used sigmoid because it's a binary classification. The employee either attritted or not.

3. We could try improving the model by re-evaluating the feature set. Maybe other features we didn't use also have predictive power, or some that we used aren't very important. Along with this, we could try feature engineering like creating interactions or grouping. We also did not attempt to tune the hyperparameters of the model like training epochs and batch size. This might lead to improvements. We could also analyze the predictions of the model using methods like a confusion matrix to see what areas the model has worse performance and try to focus on improving those specific areas. We also did not apply any methods to deal with an imbalanced class which should be helpful given our Attrition rate is about 16% and our Department has very few entries in the 'Human Resources' category.